File: 2024/typescript/day16.ts
import { type DayModule, type SolutionFactory } from './types'
import * as utils from './utils'

type Pos = [number, number]
type Path = [Pos] | [Pos, Path]
type Orientation = 'north' | 'east' | 'south' | 'west'
type State = [Pos, Orientation, Path]
type Score = number
type Field = {
    isWall: boolean
    scores: Record<Orientation, Score>
}
type Maze = Field[][]

function parse(lines: string[]): Maze {
    return lines.map(line =>
        line.split('').map(c => ({
            isWall: c === '#',
            scores: {
                north: Infinity,
                east: Infinity,
                south: Infinity,
                west: Infinity,
            },
        })),
    )
}

function left(orientation: Orientation): Orientation {
    return orientation === 'north'
        ? 'west'
        : orientation === 'west'
        ? 'south'
        : orientation === 'south'
        ? 'east'
        : 'north'
}

function right(orientation: Orientation): Orientation {
    return orientation === 'north'
        ? 'east'
        : orientation === 'east'
        ? 'south'
        : orientation === 'south'
        ? 'west'
        : 'north'
}

type Result = { score: Score; paths: Path[] }
type CancelPathPred = (previousScore: Score, currentScore: Score) => boolean
type JoinPathFn = (a: Pos, b: Path) => Path
function walk(
    maze: Maze,
    cancelCurrentPath: CancelPathPred,
    joinPath: JoinPathFn,
): Result {
    const startPos = [1, maze.length - 2] as Pos
    const [startX, startY] = startPos
    maze[startY][startX].scores.east = 0
    const [endX, endY] = [maze[0].length - 2, 1] as Pos

    function isEnd([[x, y], _orientation, _path]: State): boolean {
        return x === endX && y === endY
    }

    function tryMove([[x, y], orientation, path]: State): State | null {
        const current = maze[y][x]
        const [x2, y2] = [
            x + (orientation === 'east' ? 1 : orientation === 'west' ? -1 : 0),
            y +
                (orientation === 'south'
                    ? 1
                    : orientation === 'north'
                    ? -1
                    : 0),
        ] as Pos

        const next = maze[y2][x2]
        if (next.isWall) {
            return null
        }

        const score = current.scores[orientation] + 1
        if (cancelCurrentPath(next.scores[orientation], score)) {
            return null
        }

        next.scores[orientation] = score
        return [[x2, y2], orientation, joinPath([x2, y2], path)]
    }

    function tryTurn(
        turn: (o: Orientation) => Orientation,
        [pos, orientation, path]: State,
    ): State | null {
        const [x, y] = pos
        const current = maze[y][x]
        const nextOrientation = turn(orientation)
        const score = current.scores[orientation] + 1000
        if (cancelCurrentPath(current.scores[nextOrientation], score)) {
            return null
        }

        current.scores[nextOrientation] = score
        return [pos, nextOrientation, path]
    }

    const stack: State[] = [[startPos, 'east', [startPos]]]
    const result: Result = { score: Infinity, paths: [] }
    let state: State | undefined
    while ((state = stack.pop())) {
        let newState = tryMove(state)
        if (newState) {
            const [_, newOrientation, newPath] = newState
            if (isEnd(newState)) {
                const score = maze[endY][endX].scores[newOrientation]
                if (score < result.score) {
                    result.score = score
                    result.paths = [newPath]
                } else if (score === result.score) {
                    result.paths.push(newPath)
                }
            } else {
                stack.push(newState)
            }
        }

        if ((newState = tryTurn(left, state))) {
            stack.push(newState)
        }

        if ((newState = tryTurn(right, state))) {
            stack.push(newState)
        }
    }

    return result
}

function lowestScore(maze: Maze): number {
    return walk(
        maze,
        (previousScore, currentScore) => previousScore <= currentScore,
        (_, path) => path,
    ).score
}

function flat(path: Path): Pos[] {
    return path.length === 1 ? path : [path[0], ...flat(path[1])]
}

function posCount(maze: Maze): number {
    return new Set(
        walk(
            maze,
            (previousScore, currentScore) => previousScore < currentScore,
            (pos, path) => [pos, path],
        ).paths.flatMap(path => flat(path).map(([x, y]) => `${x},${y}`)),
    ).size
}

export const part1: Part = input => () => lowestScore(parse(input))

export const part2: Part = input => () => posCount(parse(input))

export const day = import.meta.file.match(/day(\d+)/)![1]
export const input = await utils.readInputLines(day)
part1.solution = 85432
part2.solution = 465

export const main = import.meta.main
if (main) {
    const module: Module = await import(import.meta.path)

    const testInput = utils.as_lines([
        `
###############
#.......#....E#
#.#.###.#.###.#
#.....#.#...#.#
#.###.#####.#.#
#.#.#.......#.#
#.#.#####.###.#
#...........#.#
###.#.#####.#.#
#...#.....#.#.#
#.#.#.###.#.#.#
#.....#...#.#.#
#.###.#.#.#.#.#
#S..#.....#...#
###############
`,
        `
#################
#...#...#...#..E#
#.#.#.#.#.#.#.#.#
#.#.#.#...#...#.#
#.#.#.#.###.#.#.#
#...#.#.#.....#.#
#.#.#.#.#.#####.#
#.#...#.#.#.....#
#.#.#####.#.###.#
#.#.#.......#...#
#.#.###.#####.###
#.#.#...#.....#.#
#.#.#.#####.###.#
#.#.#.........#.#
#.#.#.#########.#
#S#.............#
#################
`,
    ])

    await utils.tests(
        () =>
            utils.test_all(
                ['Test part 1', 7036, part1(testInput[0])],
                ['Test part 1', 11048, part1(testInput[1])],
                ['Test part 2', 45, part2(testInput[0])],
                ['Test part 2', 64, part2(testInput[1])],
            ),
        () => utils.test_day(module),
    )
}

type In = typeof input
type Out = number
type Module = DayModule<Out, In>
type Part = SolutionFactory<Out, In>

File: 2024/zig/src/day16.zig
const std = @import("std");
const Allocator = std.mem.Allocator;

const aoc = @import("aoc_utils");

const Turn = enum(u2) { left, right, straight };
const Direction = enum(u2) { north, east, south, west };
const Grid = aoc.Grid(Tile, u8);
const Score = u32;
const Tile = struct {
    free: bool,
    scores: [4]Score,
    pub fn init(c: u8) Tile {
        return .{
            .free = c != '#',
            .scores = [_]Score{std.math.maxInt(u32)} ** 4,
        };
    }
};
const Reindeer = struct {
    pos: Grid.Pos,
    dir: Direction,

    fn next(self: Reindeer, turn: Turn, grid: Grid) ?Reindeer {
        const d: u3 = @intFromEnum(self.dir);
        const dir: Direction = @enumFromInt(switch (turn) {
            .left => (d + 3) % 4,
            .right => (d + 1) % 4,
            .straight => d,
        });

        const pos: Grid.Pos = switch (dir) {
            .north => .{ .x = self.pos.x, .y = self.pos.y - 1 },
            .south => .{ .x = self.pos.x, .y = self.pos.y + 1 },
            .east => .{ .x = self.pos.x + 1, .y = self.pos.y },
            .west => .{ .x = self.pos.x - 1, .y = self.pos.y },
        };

        return if (grid.at(pos).free) .{ .pos = pos, .dir = dir } else null;
    }
};

const turns = std.enums.values(Turn);

const Mode = enum { min_score, field_count };

pub fn solve(comptime mode: Mode, input: []const u8, gpa: Allocator) !u32 {
    var grid: Grid = try .initMapped(input, Tile.init, gpa);
    defer grid.deinit(gpa);

    const Path = switch (mode) {
        .min_score => void,
        .field_count => std.SinglyLinkedList,
    };
    const PathNode = switch (mode) {
        .min_score => void,
        .field_count => struct { pos: Grid.Pos, node: Path.Node },
    };

    const State = switch (mode) {
        .min_score => Reindeer,
        .field_count => struct { reindeer: Reindeer, path: Path },
    };

    const Result = switch (mode) {
        .min_score => u32,
        .field_count => struct {
            score: u32,
            paths: std.ArrayList(Path),
        },
    };

    var arena: std.heap.ArenaAllocator = .init(gpa);
    defer arena.deinit();
    const gpa_arena = arena.allocator();

    var stack: std.ArrayList(State) = .empty;
    defer stack.deinit(gpa);
    const start = Grid.Pos{ .x = 1, .y = grid.height - 2 };
    grid.ptr(start).scores[@intFromEnum(Direction.east)] = 0;
    const end = Grid.Pos{ .x = grid.width - 2, .y = 1 };

    const state: State = switch (mode) {
        .min_score => .{ .pos = start, .dir = .east },
        .field_count => init: {
            var startPath: Path = .{};
            var startNode = try gpa_arena.create(PathNode);
            startNode.pos = start;
            startPath.prepend(&startNode.node);
            break :init .{
                .reindeer = .{ .pos = start, .dir = .east },
                .path = startPath,
            };
        },
    };
    try stack.append(gpa, state);

    var result: Result = switch (mode) {
        .min_score => std.math.maxInt(u32),
        .field_count => Result{
            .score = std.math.maxInt(u32),
            .paths = .empty,
        },
    };
    defer if (mode == .field_count) result.paths.deinit(gpa);

    while (stack.pop()) |s| {
        const r = if (mode == .min_score) s else s.reindeer;
        if (r.pos.x == end.x and r.pos.y == end.y) {
            const cur_score = grid.at(r.pos).scores[@intFromEnum(r.dir)];
            switch (mode) {
                .min_score => result = @min(result, cur_score),
                .field_count => {
                    if (cur_score > result.score) continue;
                    if (cur_score < result.score) {
                        result.paths.clearRetainingCapacity();
                        result.score = cur_score;
                    }
                    try result.paths.append(gpa, s.path);
                },
            }
            continue;
        }

        const current_score = grid.at(r.pos).scores[@intFromEnum(r.dir)];
        for (turns) |turn| {
            if (r.next(turn, grid)) |next| {
                const points: u32 = switch (turn) {
                    .left, .right => 1001,
                    .straight => 1,
                };
                const new_score = current_score + points;
                const next_score_ptr = &grid.ptr(next.pos).scores[@intFromEnum(next.dir)];
                const skip = switch (mode) {
                    .min_score => new_score >= next_score_ptr.*,
                    .field_count => new_score > next_score_ptr.*,
                };
                if (skip) continue;

                next_score_ptr.* = new_score;

                const next_state: State = switch (mode) {
                    .min_score => next,
                    .field_count => blk: {
                        var next_path = s.path;
                        var next_node = try gpa_arena.create(PathNode);
                        next_node.pos = next.pos;
                        next_path.prepend(&next_node.node);
                        break :blk .{ .reindeer = next, .path = next_path };
                    },
                };
                try stack.append(gpa, next_state);
            }
        }
    }

    return switch (mode) {
        .min_score => result,
        .field_count => blk: {
            var fieldSet: std.DynamicBitSetUnmanaged = try .initEmpty(gpa, grid.buf.len);
            defer fieldSet.deinit(gpa);

            for (result.paths.items) |path| {
                var path_iter = path.first;
                while (path_iter) |node| : (path_iter = node.next) {
                    const p: *PathNode = @fieldParentPtr("node", node);
                    const idx = grid.posToIndex(p.pos);
                    fieldSet.set(idx);
                }
            }
            break :blk @intCast(fieldSet.count());
        },
    };
}

pub fn part1(input: []const u8, gpa: Allocator) !Day.Result1 {
    return solve(.min_score, input, gpa);
}

pub fn part2(input: []const u8, gpa: Allocator) !Day.Result2 {
    return solve(.field_count, input, gpa);
}

pub const Day = aoc.DayInfo("16", u32, u32, 85432, 465, @This(), &.{
    .{
        .expected1 = 7036,
        .expected2 = 45,
        .input =
        \\###############
        \\#.......#....E#
        \\#.#.###.#.###.#
        \\#.....#.#...#.#
        \\#.###.#####.#.#
        \\#.#.#.......#.#
        \\#.#.#####.###.#
        \\#...........#.#
        \\###.#.#####.#.#
        \\#...#.....#.#.#
        \\#.#.#.###.#.#.#
        \\#.....#...#.#.#
        \\#.###.#.#.#.#.#
        \\#S..#.....#...#
        \\###############
        ,
    },
    .{
        .expected1 = 11048,
        .expected2 = 64,
        .input =
        \\#################
        \\#...#...#...#..E#
        \\#.#.#.#.#.#.#.#.#
        \\#.#.#.#...#...#.#
        \\#.#.#.#.###.#.#.#
        \\#...#.#.#.....#.#
        \\#.#.#.#.#.#####.#
        \\#.#...#.#.#.....#
        \\#.#.#####.#.###.#
        \\#.#.#.......#...#
        \\#.#.###.#####.###
        \\#.#.#...#.....#.#
        \\#.#.#.#####.###.#
        \\#.#.#.........#.#
        \\#.#.#.#########.#
        \\#S#.............#
        \\#################
        ,
    },
});

test "samples 1" {
    try Day.testPart1Samples();
}
test "samples 2" {
    try Day.testPart2Samples();
}
test "part 1" {
    try Day.testPart1();
}
test "part 2" {
    try Day.testPart2();
}
